lib/tools/system-calls.js

```diff
@@ -855,25 +845,43 @@ systemCallMethods.reboot = async function reboot (retries = DEFAULT_ADB_REBOOT_R
   });
 };
 
+/**
+ * @typedef {Object} rootResult
+ * @property {boolean} isRooted True if the device is rooted now
+ * @property {boolean} wasAlreadyRooted True if the device was already rooted
+ */
+
 /**
  * Switch adb server to root mode.
- *
- * @return {boolean} True of the switch was successful or false
- *                   if the switch failed.
+ * @param {boolean} unroot - Should we unroot instead of root? (default false)
+ * @return {boolean} True if the device is rooted.
  */
-systemCallMethods.root = async function root () {
+systemCallMethods.root = async function root (unroot = false) {
+  let cmd = unroot ? 'unroot' : 'root';
+  let wasAlreadyRooted = false;
   try {
-    let {stdout} = await exec(this.executable.path, ['root']);
+    let {stdout} = await exec(this.executable.path, [cmd]);
 
     // on real devices in some situations we get an error in the stdout
-    if (stdout && stdout.indexOf('adbd cannot run as root') !== -1) {
-      throw new Error(stdout.trim());
+    if (stdout) {
+      if (stdout.includes('adbd cannot run as root')) {
+        throw new Error(stdout.trim());
+      }
+      if (stdout.includes('already running as root')) {
+        wasAlreadyRooted = true;
+      }
     }
-
-    return true;
+    return {isRooted: !unroot, wasAlreadyRooted};
   } catch (err) {
-    log.warn(`Unable to root adb daemon: '${err.message}'. Continuing`);
-    return false;
+    const {stderr = ''} = err;
+    log.warn(`Unable to ${cmd} adb daemon. Original error: '${stderr}'. Continuing.`);
+    // adb 'root' can cause device to go offline, restart if so
+    if (stderr.toLowerCase().includes('closed')) {
+      log.warn(`${cmd} caused device to go offline. Restarting adb.`);
+      await this.restartAdb();
+    }
+
+    return {isRooted: unroot};
```

Comment:
I assume both `isRooted` and `wasRooted ` should be always returned

Reply:
is adb still in rooted state after restart?

Reply:
I changed it to: `isSuccessful` to indicate that the shell to `root/unroot` was successful or not and `wasAlreadyRooted` tells you if it was rooted already.

Reply:
nvm I see what you mean now....

I fixed it.